File: 2015/c/2015-01.c
#include <stdio.h>
#include <string.h>

#include "tools.h"

int part1(char data[]) {
  int output = 0;

  int i;
  for (i = 0; i < strlen(data); i++) {
    if (data[i] == '(') {
      output++;
    } else if(data[i] == ')') {
      output--;
    }

  }

  return output;
}

int part2(char data[]) {
  int output = 0;
  int i;
  int pos = 0;

  for (i = 0; i < strlen(data); i++) {
    if (data[i] == '(') {
      pos++;
    } else if (data[i] == ')') {
      pos--;
    }

    if (pos == -1) {
      output = i + 1;
      break;
    }
  }

  return output;
}

int main() {
  char *data = getDataFromFile("../data/2015-01.txt");

  printf("Advent of code 2015 #01\n\n");

  printf("----Part 1----\n");
  test("Test 1", part1("(())"),     0);
  test("Test 2", part1("()()"),     0);
  test("Test 3", part1("((("),      3);
  test("Test 4", part1("(()(()("),  3);
  test("Test 5", part1("))((((("),  3);
  test("Test 6", part1("())"),     -1);
  test("Test 7", part1("))("),     -1);
  test("Test 8", part1(")))"),     -3);
  test("Test 9", part1(")())())"), -3);
  printf("Answer: %d\n\n", part1(data));


  printf("----Part 2----\n");
  test("Test 1", part2(")"),     1);
  test("Test 2", part2("()())"), 5);
  printf("Answer: %d\n\n", part2(data));

  return 0;
}

File: 2015/js/2015-01.js
const path = require('path')
const fs   = require('fs')

const data = fs.readFileSync(path.resolve(__dirname, '../data/2015-01.txt'), 'utf8')

function part1(input) {
  let output = 0

  input.split('').forEach(step => {
    if (step == '(') {
      output++
    } else if (step == ')') {
      output--
    }
  })

  return output
}

function part2(input) {
  let pos = 0
  let output
  const data = input.split('')

  for (let i = 0; i < data.length; i++) {
    if (data[i] == '(') {
      pos++
    } else if (data[i] == ')') {
      pos--
    }

    if (pos == -1) {
      output = i + 1
      break
    }
  }

  return output
}

console.assert( part1('(())')    ==  0 )
console.assert( part1('()()')    ==  0 )
console.assert( part1('(((')     ==  3 )
console.assert( part1('(()(()(') ==  3 )
console.assert( part1('))(((((') ==  3 )
console.assert( part1('())')     == -1 )
console.assert( part1('))(')     == -1 )
console.assert( part1(')))')     == -3 )
console.assert( part1(')())())') == -3 )

console.assert( part2(')')     == 1 )
console.assert( part2('()())') == 5 )

console.log( `
  Advent of code 2015 #01
  Part1: ${part1(data)}
  Part2: ${part2(data)}
` )